### PR TITLE
backend-app-api: add shared environments

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Work towards #13082

This introduces a shared environment construct that can be used across multiple backends to simplify split backend deployments. It's pretty lightweight at the moment, only allowing for services to be added, but it's prepared for potentially adding more things in the future. It uses the same callback-based options as the one being introduced in #15720, so that you can customize the environment for each backend with options.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/backend-plugin-api/src/wiring/createSharedEnvironment.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+import { ServiceFactory, ServiceFactoryOrFunction } from '../services';
+
+/** @public */
+export interface SharedBackendEnvironmentConfig {
+  services?: ServiceFactoryOrFunction[];
+}
+
+// This type is opaque in order to allow for future API evolution with
+// cluttering the external API. For example with might want to add support
```
Comment (freben) [inline packages/backend-plugin-api/src/wiring/createSharedEnvironment.ts:25]:
```suggestion
// This type is opaque in order to allow for future API evolution without
// cluttering the external API. For example we might want to add support
```

Comment (freben) [reply packages/backend-plugin-api/src/wiring/createSharedEnvironment.ts:25]: Some `* */` and stuff below too but nothing major

File: packages/backend-plugin-api/src/wiring/createSharedEnvironment.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    if (duplicates.size > 0) {
+      const dupStr = [...duplicates].map(id => `'${id}'`).join(', ');
+      throw new Error(
+        `Duplicate service implementations provided in shared environment for ${dupStr}`,
+      );
+    }
+
+    return { version: 'v1', services } as unknown as SharedBackendEnvironment;
```
Comment (freben) [inline packages/backend-plugin-api/src/wiring/createSharedEnvironment.ts:83]: as mentioned elsewhere, i'd vote to put in the actual $$type at runtime too :) Just not to lie

File: packages/backend-defaults/src/CreateBackend.test.ts [hunk 3]
```diff
@@ -69,4 +76,91 @@ describe('createBackend', () => {
...
+          }),
+          createServiceFactory({
+            service: barServiceRef,
+            deps: {},
+            async factory() {
+              return 'bar-env';
+            },
+          }),
+        ],
+      })(),
```
Comment (freben) [inline packages/backend-defaults/src/CreateBackend.test.ts:133]: So here we're not going to do the thing where createSharedEnvironment only returns a function if its input was a function? :)

**Review by freben — APPROVED** [review]
Looks good, just optional thoughts

File: packages/backend-defaults/src/CreateBackend.test.ts (hunk 3 above)
Comment (Rugvip) [reply packages/backend-defaults/src/CreateBackend.test.ts:133]: Nah, always a function right? Feeling it makes sense even for this somewhat more internal helper. Makes it easier to keep backwards compatibility when using this in an internally shared package too.

File: packages/backend-plugin-api/src/wiring/createSharedEnvironment.ts (hunk 2 above)
Comment (Rugvip) [reply packages/backend-plugin-api/src/wiring/createSharedEnvironment.ts:83]: Yep, done, and made sure we've got type safety here too.

**Pull request merged**